### PR TITLE
Spreadsheed Formula Functions

### PR DESCRIPTION
**Pull request by zgordan-vv**
Addresses part of #337.

This change is [Reviewable](https://reviewable.io/reviews/unidoc/unioffice/345)

### CODE REVIEW

File: spreadsheet/formula/fnstatistical.go [hunk 1]
```diff
@@ -121,6 +122,37 @@ func CountBlank(args []Result) Result {
 	return MakeNumberResult(count(args, countEmpty))
 }
 
+// CountIf implements the COUNTIF function.
+func CountIf(args []Result) Result {
+	// COUNT and COUNTA don't require arguments, COUNTBLANK does
```
Comment (gunnsth) [inline spreadsheet/formula/fnstatistical.go]: Is this used for COUNT/COUNTA/COUNTBLANK too? Not just COUNTIF?

File: spreadsheet/formula/functions_test.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	"strings"
+	"testing"
+
+	"github.com/unidoc/unioffice/measurement"
+	"github.com/unidoc/unioffice/schema/soo/sml"
+	"github.com/unidoc/unioffice/spreadsheet"
+	"github.com/unidoc/unioffice/spreadsheet/formula"
+)
+
+type testStruct struct {
```
Comment (gunnsth) [inline spreadsheet/formula/functions_test.go]: Add a comment on what Inp and Exp are?  Inp presumably Input?  And Exp expected?
What format is the Expected string?  Are the results generic or are they for a specific workbook?

File: spreadsheet/formula/functions_test.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+	sciStyle := ss.StyleSheet.AddCellStyle()
+	sciStyle.SetNumberFormat("##0.00E+00")
+	sheet.Cell("A24").SetStyle(sciStyle)
+
+	incorrectStyle := ss.StyleSheet.AddCellStyle()
+	incorrectStyle.SetNumberFormat("incorrect style")
+	sheet.Cell("A25").SetStyle(incorrectStyle)
+
+// for testing alignments ("prefix" function)
```
Comment (gunnsth) [inline spreadsheet/formula/functions_test.go]: indent

File: spreadsheet/formula/functions_test.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	sheet.Cell("B3").SetString("value3")
+
+	ctx := sheet.FormulaContext()
+
+	runTests(t, ctx, td)
+}
+
+func TestColumn(t *testing.T) {
+	td := []testStruct{
+		{`=COLUMN(A1)`, `1 ResultTypeNumber`},
```
Comment (gunnsth) [inline spreadsheet/formula/functions_test.go]: Are these default results? Seems like the test is run on a freshly created empty file

File: spreadsheet/formula/functions_test.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	Inp string
+	Exp string
+}
+
+func runTests(t *testing.T, ctx formula.Context, td []testStruct) {
+	ev := formula.NewEvaluator()
+	for _, tc := range td {
+		t.Run(tc.Inp, func(t *testing.T) {
+			p := formula.Parse(strings.NewReader(tc.Inp))
+			if p == nil {
```
Comment (gunnsth) [inline spreadsheet/formula/functions_test.go]: Might be nice to use testify require.  require.NotNil(t, p)
etc

File: spreadsheet/formula/functions_test.go (hunk 2 above)
Comment (gunnsth) [reply spreadsheet/formula/functions_test.go]: After looking more, seems like the "Exp" is the value and the type. Would be good to mention that in a comment.
Also rather use Input and Expected, no need to abbreviate those.
// Input is an input formula string.
// Expected is the expected output of the formula as a string of format: "value type"

It's also worth mentioning that the Exp depends on the workbook that is being worked with, not just the Input.

File: spreadsheet/formula/invalidreferencecontext.go [hunk 6]
```diff
@@ -30,3 +30,27 @@ func (i *ivr) Sheet(name string) Context {
 func (i *ivr) SetOffset(col, row uint32) {
 
 }
+
+func (i *ivr) GetFormat(cellRef string) string {
```
Comment (gunnsth) [inline spreadsheet/formula/invalidreferencecontext.go]: godoc, are these supposed to be empty?

File: spreadsheet/read.go [hunk 7]
```diff
@@ -67,5 +68,11 @@ func Open(filename string) (*Workbook, error) {
 	if err != nil {
 		return nil, fmt.Errorf("error opening %s: %s", filename, err)
 	}
-	return Read(f, fi.Size())
+	wb, err := Read(f, fi.Size())
+	if err != nil {
+		return nil, err
+	}
+	dir, _ := filepath.Abs(filepath.Dir(filename))
+	wb.filename = dir + "/" + filename
```
Comment (gunnsth) [inline spreadsheet/read.go]: filepath.Join

File: spreadsheet/workbook.go [hunk 8]
```diff
@@ -744,3 +745,7 @@ func (wb *Workbook) RemoveCalcChain() {
 		}
 	}
 }
+
+func (wb *Workbook) GetFilename() string {
```
Comment (gunnsth) [inline spreadsheet/workbook.go]: godoc

File: spreadsheet/formula/functions_test.go (hunk 2 above)
Comment (zgordan-vv) [reply spreadsheet/formula/functions_test.go]: Just copied this struct with Inp and Exp from spreadsheet/formula/eval_test.go , will fix. Maybe it worths to be fixed in eval_test.go too.

File: spreadsheet/formula/functions_test.go (hunk 4 above)
Comment (zgordan-vv) [reply spreadsheet/formula/functions_test.go]: yes, column A is always 1, column B is always 2, no need to modify freshly created empty file

File: spreadsheet/formula/invalidreferencecontext.go (hunk 6 above)
Comment (zgordan-vv) [reply spreadsheet/formula/invalidreferencecontext.go]: Yes, they are supposed to be empty and present there only to implement the interface. This type implements interface Context of formula package and all its methods are documented in the definition of the interface (spreadsheet/formula/context.go). Do we really need to godoc it?

File: spreadsheet/formula/fnstatistical.go (hunk 1 above)
Comment (zgordan-vv) [reply spreadsheet/formula/fnstatistical.go]: removed, this comment was occasionally copypasted

File: spreadsheet/formula/functions_test.go (hunk 5 above)
Comment (zgordan-vv) [reply spreadsheet/formula/functions_test.go]: golang standard testing package doesn't contain assert or require methods, do you consider to use github.com/stretchr/testify/require ?

File: spreadsheet/formula/functions_test.go (hunk 2 above)
Comment (zgordan-vv) [reply spreadsheet/formula/functions_test.go]: fixed

File: spreadsheet/formula/functions_test.go (hunk 3 above)
Comment (zgordan-vv) [reply spreadsheet/formula/functions_test.go]: fixed

File: spreadsheet/read.go (hunk 7 above)
Comment (zgordan-vv) [reply spreadsheet/read.go]: fixed

File: spreadsheet/workbook.go (hunk 8 above)
Comment (zgordan-vv) [reply spreadsheet/workbook.go]: fixed

File: spreadsheet/formula/context.go [hunk 9]
```diff
@@ -17,6 +17,24 @@ type Context interface {
...
+	// GetFilename returns the full filename of the context's Workbook
+	GetFilename() string
+
+	// GetWidth returns a worksheet's column width
+	GetWidth(colIdx int) float64
+
+	// GetFormat returns a cell's format
+	GetFormat(cellRef string) string
+
+	// GetFormat returns cell's label prefix dependent on cell horizontal alignment
```
Comment (gunnsth) [inline spreadsheet/formula/context.go]: GetLabelPrefix...
Each comment sentence should end with a dot also as per godoc convention. I.e.
https://golang.org/doc/effective_go.html#commentary
```
Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.
```

File: spreadsheet/formula/context.go [hunk 10]
```diff
@@ -17,6 +17,24 @@ type Context interface {
...
+	GetLabelPrefix(cellRef string) string
+
+	// GetFormat returns if cell is protected
+	GetLocked(cellRef string) bool
+
+	// GetFormat returns sets cell's protected attribute
```
Comment (gunnsth) [inline spreadsheet/formula/context.go]: Check the other comments too

File: spreadsheet/formula/fnindex.go [hunk 11]
```diff
@@ -24,6 +27,54 @@ func init() {
 	RegisterFunction("TRANSPOSE", Transpose)
 }
 
+// Choose implements the Excel CHOOSE function
```
Comment (gunnsth) [inline spreadsheet/formula/fnindex.go]: sentence should end with dot

File: spreadsheet/formula/invalidreferencecontext.go (hunk 6 above)
Comment (gunnsth) [reply spreadsheet/formula/invalidreferencecontext.go]: If it's more specific.  For example here.
// GetFormat returns an empty string for the invalid reference context.
It's better to keep the style consistent.
If golint complains about it, we should do it (most often at least).

File: spreadsheet/formula/context.go (hunk 9 above)
Comment (zgordan-vv) [reply spreadsheet/formula/context.go]: fixed

File: spreadsheet/formula/context.go (hunk 10 above)
Comment (zgordan-vv) [reply spreadsheet/formula/context.go]: checked

File: spreadsheet/formula/fnindex.go (hunk 11 above)
Comment (zgordan-vv) [reply spreadsheet/formula/fnindex.go]: fixed

File: spreadsheet/formula/invalidreferencecontext.go (hunk 6 above)
Comment (zgordan-vv) [reply spreadsheet/formula/invalidreferencecontext.go]: done

File: spreadsheet/formula/functions_test.go (hunk 5 above)
Comment (gunnsth) [reply spreadsheet/formula/functions_test.go]: Yes I was referencing that one.

Comment (gunnsth) [reply spreadsheet/formula/functions_test.go]: Doesn't matter much in this specific case here, but can be nice in longer test cases with many
comparisons etc.

**Comment by zgordan-vv** [comment]
> Looks good. @zgordan-vv Is this PR ready for merging, or are you still adding to it?

I can start another PR to not to delay things

**Pull request merged**